src/humanloop/eval_utils/run.py

```diff
@@ -0,0 +1,739 @@
+            Flow.model_validate(version)
+        except ValidationError:
+            flow_version = {"attributes": version}
+            file_dict = {**file_, **flow_version}
+        hl_file = client.flows.upsert(**file_dict)  # type: ignore
+
+    elif type_ == "prompt":
+        try:
+            Prompt.model_validate(version)
+        except ValidationError as error_:
+            logger.error(msg="Invalid Prompt `version` in your `file` request. \n\nValidation error: \n)")
+            raise error_
+        try:
+            hl_file = client.prompts.upsert(**file_dict)  # type: ignore
+        except ApiError as error_:
+            raise error_
+
+    elif type_ == "tool":
+        try:
+            Tool.model_validate(version)
+        except ValidationError as error_:
+            logger.error(msg="Invalid Tool `version` in your `file` request. \n\nValidation error: \n)")
+            raise error_
+        hl_file = client.tools.upsert(**file_dict)  # type: ignore
+
+    elif type_ == "evaluator":
+        hl_file = client.evaluators.upsert(**file_dict)  # type: ignore
+
+    else:
+        raise NotImplementedError(f"Unsupported File type: {type_}")
+
+    # Upsert the Dataset
+    hl_dataset = client.datasets.upsert(**dataset)
+    hl_dataset = client.datasets.get(id=hl_dataset.id, include_datapoints=True)
+
+    # Upsert the local Evaluators; other Evaluators are just referenced by `path` or `id`
+    local_evaluators: List[Evaluator] = []
+    if evaluators:
+        for evaluator in evaluators:
+            # If a callable is provided for an Evaluator, we treat it as External
+            eval_function = evaluator.get("callable")
+            if eval_function is not None:
+                # TODO: support the case where `file` logs generated on Humanloop but Evaluator logs generated locally
+                if function_ is None:
+                    raise ValueError(
+                        f"Local Evaluators are only supported when generating Logs locally using your {type_}'s `callable`. Please provide a `callable` for your file in order to run Evaluators locally."
+                    )
+                local_evaluators.append(evaluator)
+                spec = ExternalEvaluator(
+                    arguments_type=evaluator["args_type"],
+                    return_type=evaluator["return_type"],
+                    attributes={"code": inspect.getsource(eval_function)},
+                    evaluator_type="external",
+                )
+                client.evaluators.upsert(
+                    id=evaluator.get("id"),
+                    path=evaluator.get("path"),
+                    spec=spec,
+                )
+    function_ = typing.cast(Callable, function_)
+
+    # Validate signature of the called function
+    function_signature = inspect.signature(function_)
+    parameter_names = list(function_signature.parameters.keys())
+    if parameter_names != ["inputs", "messages"] and parameter_names != ["inputs"]:
```

Comment:
This seems wrong - the way function used to be called is:

function(**inputs, messages)

So inputs are spread

Reply:
Ack fixed